README.md
# tarantool-perl
Perl driver for Tarantool 1.6.  
Fully featured client to new versions (>=1.6) of [tarantool](http://tarantool.org) database, providing methods for all commonly used database query types.

Currently in process of uploading on CPAN.  

#### Features
  - Async queries (event-driven development)
  - Sync queries
  - Ability to use spaces and indexes names (schema loading) in following queries
  - SELECT, INSERT, DELETE, REPLACE, UPDATE, UPSERT, LUA functions call  
   
#### Usage  
Since tarantool database has two completely different versions (1.5 and 1.6), tarantool-perl client is created to be universal, which is able to handle both versions.

##### Tarantool version 1.5
Methods are located in namespace **DR::Tarantool::\***

##### Tarantool version >= 1.6
Methods are located in namespace **DR::Tarantool::MsgPack::\*** .  Some examples are listed below.
First, let's use synchronous client to get connection to database:
```perl
my $connection = DR::Tarantool::MsgPack::SyncClient->connect(
    host => $host,
    port => $port);
```
Using this object one can call various methods like:
```perl
my $data = $connection->select('space_name', 'index_name', [ 22 ]);
```
to fetch data from space *space_name* using index *index_name*.  

To be honest, *sync* client is a simple wrapper for *async* one, so the core of the application is projected to be *async*. The following examples are to demonstrate the way async methods should be called (here [AnyEvent](https://metacpan.org/pod/AnyEvent) library is used for event-loop implementation):
```perl
use AnyEvent;
use Data::Dumper;

my $cv = AnyEvent->condvar();
DR::Tarantool::MsgPack::AsyncClient->connect(
    host => $host,
    port => $port,
    sub {
        shift->select(
            'space_name', 
            'index_name',
            [22],
            sub { print Dumper $_[1]->raw; $cv->send }
        );
    });
# ... your code here
$cv->recv();
```
Here we pass one more argument to `connect()` method, which is callback to be invocked when the connection has actually been established. In this callback we in turn call `select()` method with its own callback, which dumps the structure *select* returned and stops event-loop.

The other methods are to be invoked in a similar way, the functions prototypes and usage examples can be found in perldoc.

### Version
0.60

### Tech

Main dependencies for this package are the following perl modules:

* [AnyEvent] - event-loop implementation
* [Coro] - coroutines

### Installation

Until the module is released on CPAN in order to install tarantool-perl its sources can be downloaded from here and compiled (lowlevel parts are written with XS).

```sh
$ perl Makefile.PL
$ make && make install
```

### Development

Want to contribute? Great! It's github, open issues, send pull requests :)

### TODO

 - Make the way the errors are delivered from all methods clear

### License

Artistic


Changes
tarantool-perl repo (0.60)

  * Fixed: new data types in Tarantool, not to fall while schema downloading, closes #22
  * Fixed: soft schema reload, closes #24

 -- Pavel Limorenko <pavel@limorenko.com> Sub, 20 Aug 2017 19:42:00 +0400


tarantool-perl repo (0.59)

  * Fixed: correctly close tarantool connection on fatal error (timeout, error from tarantool)

 -- Pavel Limorenko <p.limorenko@corp.mail.ru> Tue, 14 Feb 2017 20:10:20 +0400


tarantool-perl repo (0.58)

  * Fixed 'die' on unexpected reply from tarantool

 -- Pavel Limorenko <p.limorenko@corp.mail.ru> Mon, 13 Feb 2017 11:39:19 +0400


tarantool-perl repo (0.57)

  * Added lazy=>1 connect, all connection issues are hidden from user, closes: #21
  * Added {on}{connected} user callback support to track if the connection has been established
  * Fixed bug with number of connect attempts when connection has previously been established and lost then

 -- Pavel Limorenko <p.limorenko@corp.mail.ru> Sun, 11 Dec 2016 20:01:31 +0400


tarantool-perl repo (0.56)

  * Fixed memleak when performing request, closes: #20
  * Implemented request_timeout, closes: #19
  * Made connection settings accessors to be available for AsyncClient object
  * Made reconnect() method for AsyncClient object (to be used with disconnect())

 -- Pavel Limorenko <p.limorenko@corp.mail.ru> Mon, 31 Oct 2016 12:19:26 +0400


tarantool-perl repo (0.55)

  * Made socket read to be handled only when the request has been sent, closes: #18

 -- Pavel Limorenko <p.limorenko@corp.mail.ru> Fri, 21 Oct 2016 19:29:26 +0400


tarantool-perl repo (0.54)

  * Fixed auth after reconnect, closes: #17

 -- Pavel Limorenko <p.limorenko@corp.mail.ru> Fri, 9 Sep 2016 12:04:43 +0400


tarantool-perl repo (0.53)

  * Added possibility to specify callbacks for connect events like error, connfail, closes: #12
  * Added possibility to specify connect timeout and number of connect attempts, closes: #14

 -- Pavel Limorenko <p.limorenko@corp.mail.ru> Mon, 29 Aug 2016 15:08:36 +0400

tarantool-perl repo (0.52)

  * Fixed: sednings strings that looks_like_number() as strings

 -- Pavel Limorenko <p.limorenko@corp.mail.ru> Mon, 1 Aug 2016 19:37:56 +0400

tarantool-perl repo (0.51)

  * Fixed: schema loading access problem

 -- Pavel Limorenko <p.limorenko@corp.mail.ru> Wed, 27 Jul 2016 12:05:11 +0400

tarantool-perl repo (0.50)

  * Fixed: no to get frozen on query fails because of croaks

 -- Pavel Limorenko <p.limorenko@corp.mail.ru> Wed, 27 Jul 2016 11:06:37 +0400

tarantool-perl repo (0.49)

  * Fixed: correct handling of two indexes for one space, closes: #9

 -- Pavel Limorenko <p.limorenko@corp.mail.ru> Wed, 29 Jun 2016 11:42:57 +0400


tarantool-perl repo (0.48)

  * Fixed: correct handling of disconnection, closes: #10
  * Added 'disconnect' method for classes AnyncClient and SyncClient

 -- Pavel Limorenko <p.limorenko@corp.mail.ru> Mon, 27 Jun 2016 15:13:28 +0400


tarantool-perl repo (0.47)

  * Fixed: correct handling of connection error, closes: #8

 -- Pavel Limorenko <p.limorenko@corp.mail.ru> Mon, 4 Apr 2016 11:02:51 +0400


tarantool-perl repo (0.46)

  * Added: support of 'upsert' method + tests, closes: #1
  * Fixed: call of 'update' method with some extra field number

 -- Pavel Limorenko <p.limorenko@corp.mail.ru> Thu, 3 Mar 2016 12:30:27 +0400


tarantool-perl repo (0.45)

  * Fixed: now call_lua() works with multiple arguments
  * Added: automatic schema loading + tests, closes: #2
  * Fixed: test llclient - in error from tarantool space number is
    now written without '#', so test now respects it

 -- Pavel Limorenko <p.limorenko@corp.mail.ru> Thu, 3 Mar 2016 12:30:27 +0400



lib/DR/Tarantool/Constants.pm
use utf8;
use strict;
use warnings;

package DR::Tarantool::Constants;
use Carp;

my %ERROR_CODES = ( # from tarantool sources https://github.com/tarantool/tarantool/
    0 => "ER_UNKNOWN",
    1 => "ER_ILLEGAL_PARAMS",
    2 => "ER_MEMORY_ISSUE",
    3 => "ER_TUPLE_FOUND",
    4 => "ER_TUPLE_NOT_FOUND",
    5 => "ER_UNSUPPORTED",
    6 => "ER_NONMASTER",
    7 => "ER_READONLY",
    8 => "ER_INJECTION",
    9 => "ER_CREATE_SPACE",
   10 => "ER_SPACE_EXISTS",
   11 => "ER_DROP_SPACE",
   12 => "ER_ALTER_SPACE",
   13 => "ER_INDEX_TYPE",
   14 => "ER_MODIFY_INDEX",
   15 => "ER_LAST_DROP",
   16 => "ER_TUPLE_FORMAT_LIMIT",
   17 => "ER_DROP_PRIMARY_KEY",
   18 => "ER_KEY_PART_TYPE",
   19 => "ER_EXACT_MATCH",
   20 => "ER_INVALID_MSGPACK",
   21 => "ER_PROC_RET",
   22 => "ER_TUPLE_NOT_ARRAY",
   23 => "ER_FIELD_TYPE",
   24 => "ER_FIELD_TYPE_MISMATCH",
   25 => "ER_SPLICE",
   26 => "ER_ARG_TYPE",
   27 => "ER_TUPLE_IS_TOO_LONG",
   28 => "ER_UNKNOWN_UPDATE_OP",
   29 => "ER_UPDATE_FIELD",
   30 => "ER_FIBER_STACK",
   31 => "ER_KEY_PART_COUNT",
   32 => "ER_PROC_LUA",
   33 => "ER_NO_SUCH_PROC",
   34 => "ER_NO_SUCH_TRIGGER",
   35 => "ER_NO_SUCH_INDEX",
   36 => "ER_NO_SUCH_SPACE",
   37 => "ER_NO_SUCH_FIELD",
   38 => "ER_SPACE_FIELD_COUNT",
   39 => "ER_INDEX_FIELD_COUNT",
   40 => "ER_WAL_IO",
   41 => "ER_MORE_THAN_ONE_TUPLE",
   42 => "ER_ACCESS_DENIED",
   43 => "ER_CREATE_USER",
   44 => "ER_DROP_USER",
   45 => "ER_NO_SUCH_USER",
   46 => "ER_USER_EXISTS",
   47 => "ER_PASSWORD_MISMATCH",
   48 => "ER_UNKNOWN_REQUEST_TYPE",
   49 => "ER_UNKNOWN_SCHEMA_OBJECT",
   50 => "ER_CREATE_FUNCTION",
   51 => "ER_NO_SUCH_FUNCTION",
   52 => "ER_FUNCTION_EXISTS",
   53 => "ER_FUNCTION_ACCESS_DENIED",
   54 => "ER_FUNCTION_MAX",
   55 => "ER_SPACE_ACCESS_DENIED",
   56 => "ER_USER_MAX",
   57 => "ER_NO_SUCH_ENGINE",
   58 => "ER_RELOAD_CFG",
   59 => "ER_CFG",
   60 => "ER_SOPHIA",
   62 => "ER_UNKNOWN_SERVER",
   63 => "ER_CLUSTER_ID_MISMATCH",
   64 => "ER_INVALID_UUID",
   65 => "ER_CLUSTER_ID_IS_RO",
   66 => "ER_RESERVED66",
   67 => "ER_SERVER_ID_IS_RESERVED",
   68 => "ER_INVALID_ORDER",
   69 => "ER_MISSING_REQUEST_FIELD",
   70 => "ER_IDENTIFIER",
   71 => "ER_DROP_FUNCTION",
   72 => "ER_ITERATOR_TYPE",
   73 => "ER_REPLICA_MAX",
   74 => "ER_INVALID_XLOG",
   75 => "ER_INVALID_XLOG_NAME",
   76 => "ER_INVALID_XLOG_ORDER",
   77 => "ER_NO_CONNECTION",
   78 => "ER_TIMEOUT",
   79 => "ER_ACTIVE_TRANSACTION",
   80 => "ER_NO_ACTIVE_TRANSACTION",
   81 => "ER_CROSS_ENGINE_TRANSACTION",
   82 => "ER_NO_SUCH_ROLE",
   83 => "ER_ROLE_EXISTS",
   84 => "ER_CREATE_ROLE",
   85 => "ER_INDEX_EXISTS",
   87 => "ER_ROLE_LOOP",
   88 => "ER_GRANT",
   89 => "ER_PRIV_GRANTED",
   90 => "ER_ROLE_GRANTED",
   91 => "ER_PRIV_NOT_GRANTED",
   92 => "ER_ROLE_NOT_GRANTED",
   93 => "ER_MISSING_SNAPSHOT",
   94 => "ER_CANT_UPDATE_PRIMARY_KEY",
   95 => "ER_UPDATE_INTEGER_OVERFLOW",
   96 => "ER_GUEST_USER_PASSWORD",
   97 => "ER_TRANSACTION_CONFLICT",
   98 => "ER_UNSUPPORTED_ROLE_PRIV",
   99 => "ER_LOAD_FUNCTION",
  100 => "ER_FUNCTION_LANGUAGE",
  101 => "ER_RTREE_RECT",
  102 => "ER_PROC_C",
  103 => "ER_UNKNOWN_RTREE_INDEX_DISTANCE_TYPE",
  104 => "ER_PROTOCOL",
  105 => "ER_UPSERT_UNIQUE_SECONDARY_KEY",
  106 => "ER_WRONG_INDEX_RECORD",
  107 => "ER_WRONG_INDEX_PARTS",
  108 => "ER_WRONG_INDEX_OPTIONS",
  109 => "ER_WRONG_SCHEMA_VERSION",
  110 => "ER_SLAB_ALLOC_MAX",
  111 => "ER_WRONG_SPACE_OPTIONS",
  112 => "ER_UNSUPPORTED_INDEX_FEATURE",
  113 => "ER_VIEW_IS_RO",
);

my %service_spaces_no = (
  _vspace => 281,
  _vindex => 289,
);

sub get_space_no {
    my $space_name = shift;
    return unless exists $service_spaces_no{$space_name};
    return $service_spaces_no{$space_name};
}

sub get_error_name {
    my $error_num  = shift;
    $error_num -= 2 ** 15;

    my $error_name = $ERROR_CODES{$error_num};
    unless ( defined $error_name ) {
        return;
    }
    return $error_name;
}

1;
